core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -197,10 +196,13 @@ class TransactionStateManager(brokerId: Int,
     }, delay = config.removeExpiredTransactionalIdsIntervalMs, period = config.removeExpiredTransactionalIdsIntervalMs)
   }
 
-  def getTransactionState(transactionalId: String) = getAndMaybeAddTransactionState(transactionalId, None)
+  def getTransactionState(transactionalId: String): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]] =
+    getAndMaybeAddTransactionState(transactionalId, None)
 
-  def putTransactionStateIfNotExists(transactionalId: String, txnMetadata: TransactionMetadata) =
+  def putTransactionStateIfNotExists(transactionalId: String,
+                                     txnMetadata: TransactionMetadata): Either[Errors, CoordinatorEpochAndTxnMetadata] =
     getAndMaybeAddTransactionState(transactionalId, Some(txnMetadata))
+      .right.map(_.getOrElse(throw new IllegalStateException("Expected non-empty transaction metadata returned")))
```

Comment:
nit: `Expected non-empty transaction metadata returned` reads a bit unclear. Were you trying to say `Unexpected empty transaction metadata returned while putting $txnMetadata`?